### ISSUE TITLE
FastAPI and uvicorn dependencies are not installed in the virtual environment

### ISSUE THREAD

**Issue by panicoenlaxbox** [microsoft/Oryx#1400]
### Bug Report

- At what date and time did you most recently experience the problem?

05/28/2022

- Where did you experience the problem? E.g. Azure Web Apps, Azure Functions, Azure Container Registry, or offline use.

Azure Web App

- What happened?

Some dependencies are not installed in the virtual environment.

- What did you expect or want to happen?

That the deployment would be successful

- How can we reproduce it?

I have a very simple application with this code:

```python
import uvicorn
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def read_main():
    return {"msg": "Hello World"}


if __name__ == "__main__":
    uvicorn.run(app, host="localhost", port=8000)
```

I start it locally with `python main.py` or `uvicorn main:app --host localhost --port 8000`. I have just followed the getting started from https://fastapi.tiangolo.com/#example

`requirements.txt` has the following content (it has been created with `pipenv requirements`) and it's the project's root folder.

```
-i https://pypi.org/simple
anyio==3.6.1
asgiref==3.5.2
certifi==2022.5.18.1
charset-normalizer==2.0.12
click==8.1.3
colorama==0.4.4
cowsay==4.0
fastapi==0.78.0
h11==0.13.0
httptools==0.4.0
idna==3.3
pydantic==1.9.1
python-dotenv==0.20.0
pyyaml==6.0
requests==2.27.1
sniffio==1.2.0
starlette==0.19.1
typing-extensions==4.2.0
urllib3==1.26.9
uvicorn[standard]==0.17.6
watchgod==0.8.2
websockets==10.3
```

During the deployment, `output.tar.gz` is created but it lacks some dependencies (I'm downloading it from Kudu). Moreover, I can see it in the logs of the App Service > Deployment Center blade in the portal that it seems that some dependencies are already installed (however they are not documented at https://github.com/Azure-App-Service/python#components)

```
Python Version: /tmp/oryx/platforms/python/3.9.7/bin/python3.9
Creating directory for command manifest file if it doesnot exist
Removing existing manifest file
Python Virtual Environment: antenv
Creating virtual environment...
Activating virtual environment...
Running pip install...
[11:38:00+0000] Collecting anyio==3.6.1
[11:38:00+0000]   Using cached anyio-3.6.1-py3-none-any.whl (80 kB)
[11:38:00+0000] Collecting asgiref==3.5.2
[11:38:00+0000]   Using cached asgiref-3.5.2-py3-none-any.whl (22 kB)
[11:38:00+0000] Collecting certifi==2022.5.18.1
[11:38:00+0000]   Using cached certifi-2022.5.18.1-py3-none-any.whl (155 kB)
[11:38:00+0000] Collecting charset-normalizer==2.0.12
[11:38:00+0000]   Using cached charset_normalizer-2.0.12-py3-none-any.whl (39 kB)
[11:38:01+0000] Collecting click==8.1.3
[11:38:01+0000]   Using cached click-8.1.3-py3-none-any.whl (96 kB)
[11:38:01+0000] Collecting colorama==0.4.4
[11:38:01+0000]   Using cached colorama-0.4.4-py2.py3-none-any.whl (16 kB)
[11:38:01+0000] Collecting cowsay==4.0
[11:38:01+0000]   Using cached cowsay-4.0-py2.py3-none-any.whl (24 kB)
[11:38:01+0000] Requirement already satisfied: fastapi==0.78.0 in /home/.local/lib/python3.9/site-packages/fastapi-0.78.0-py3.9.egg (from -r requirements.txt (line 9)) (0.78.0)
[11:38:01+0000] Collecting h11==0.13.0
[11:38:01+0000]   Using cached h11-0.13.0-py3-none-any.whl (58 kB)
[11:38:01+0000] Collecting httptools==0.4.0
[11:38:02+0000]   Using cached httptools-0.4.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (433 kB)
[11:38:02+0000] Collecting idna==3.3
[11:38:02+0000]   Using cached idna-3.3-py3-none-any.whl (61 kB)
[11:38:02+0000] Collecting pydantic==1.9.1
[11:38:03+0000]   Using cached pydantic-1.9.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.4 MB)
[11:38:03+0000] Collecting python-dotenv==0.20.0
[11:38:03+0000]   Using cached python_dotenv-0.20.0-py3-none-any.whl (17 kB)
[11:38:04+0000] Collecting pyyaml==6.0
[11:38:04+0000]   Using cached PyYAML-6.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (661 kB)
[11:38:04+0000] Requirement already satisfied: requests==2.27.1 in /home/.local/lib/python3.9/site-packages/requests-2.27.1-py3.9.egg (from -r requirements.txt (line 16)) (2.27.1)
[11:38:04+0000] Collecting sniffio==1.2.0
[11:38:04+0000]   Using cached sniffio-1.2.0-py3-none-any.whl (10 kB)
[11:38:04+0000] Collecting starlette==0.19.1
[11:38:04+0000]   Using cached starlette-0.19.1-py3-none-any.whl (63 kB)
[11:38:04+0000] Collecting typing-extensions==4.2.0
[11:38:04+0000]   Using cached typing_extensions-4.2.0-py3-none-any.whl (24 kB)
[11:38:05+0000] Collecting urllib3==1.26.9
[11:38:05+0000]   Using cached urllib3-1.26.9-py2.py3-none-any.whl (138 kB)
[11:38:05+0000] Requirement already satisfied: uvicorn[standard]==0.17.6 in /home/.local/lib/python3.9/site-packages/uvicorn-0.17.6-py3.9.egg (from -r requirements.txt (line 21)) (0.17.6)
[11:38:05+0000] Collecting watchgod==0.8.2
[11:38:05+0000]   Using cached watchgod-0.8.2-py3-none-any.whl (12 kB)
[11:38:05+0000] Requirement already satisfied: websockets==10.3 in /home/.local/lib/python3.9/site-packages/websockets-10.3-py3.9-linux-x86_64.egg (from -r requirements.txt (line 23)) (10.3)
[11:38:06+0000] Collecting uvloop!=0.15.0,!=0.15.1,>=0.14.0
[11:38:06+0000]   Using cached uvloop-0.16.0-cp39-cp39-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (4.3 MB)
[11:38:07+0000] Installing collected packages: sniffio, idna, typing-extensions, h11, click, asgiref, anyio, watchgod, uvloop, urllib3, starlette, pyyaml, python-dotenv, pydantic, httptools, charset-normalizer, certifi, cowsay, colorama
[11:38:12+0000] Successfully installed anyio-3.6.1 asgiref-3.5.2 certifi-2022.5.18.1 charset-normalizer-2.0.12 click-8.1.3 colorama-0.4.4 cowsay-4.0 h11-0.13.0 httptools-0.4.0 idna-3.3 pydantic-1.9.1 python-dotenv-0.20.0 pyyaml-6.0 sniffio-1.2.0 starlette-0.19.1 typing-extensions-4.2.0 urllib3-1.26.9 uvloop-0.16.0 watchgod-0.8.2
```

The problem is with FastAPI and uvicorn.

I have verified the current location during the deployment with a `POST_BUILD_COMMAND`.

```
Package            Version     Location                                                                        Installer
------------------ ----------- ------------------------------------------------------------------------------- ---------
anyio              3.6.1       /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
asgiref            3.5.2       /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
certifi            2022.5.18.1 /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
charset-normalizer 2.0.12      /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
click              8.1.3       /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
colorama           0.4.4       /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
cowsay             4.0         /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
fastapi            0.78.0      /home/.local/lib/python3.9/site-packages/fastapi-0.78.0-py3.9.egg
h11                0.13.0      /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
httptools          0.4.0       /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
idna               3.3         /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
other-project      1.0.0       /home/.local/lib/python3.9/site-packages/other_project-1.0.0-py3.9.egg
pip                21.2.3      /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
pydantic           1.9.1       /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
python-dotenv      0.20.0      /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
PyYAML             6.0         /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
requests           2.27.1      /home/.local/lib/python3.9/site-packages/requests-2.27.1-py3.9.egg
setuptools         57.4.0      /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
sniffio            1.2.0       /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
starlette          0.19.1      /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
typing_extensions  4.2.0       /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
urllib3            1.26.9      /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
uvicorn            0.17.6      /home/.local/lib/python3.9/site-packages/uvicorn-0.17.6-py3.9.egg
uvloop             0.16.0      /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
watchgod           0.8.2       /tmp/8da409a921c1dd1/antenv/lib/python3.9/site-packages                         pip
websockets         10.3        /home/.local/lib/python3.9/site-packages/websockets-10.3-py3.9-linux-x86_64.egg
wheel              0.37.0      /tmp/oryx/platforms/python/3.9.7/lib/python3.9/site-packages                    pip
Finished executing post-build command.
Preparing output...
```

Anyway, deployment succeed, but the file `output.tar.gz` that is deployed to the container does not include these dependencies in `antenv\lib\python3.9\site-packages` . So when I try to start  (Startup command) with `uvicorn main:app --host localhost --port 8000` or `gunicorn -w 4 -k uvicorn.workers.UvicornWorker main:app` or something similar, it's always the same error message: `[ModuleNotFoundError: No module named 'uvicorn'](https://stackoverflow.com/questions/66275437/fastapi-modulenotfounderror-no-module-named-uvicorn)`

With a custom script (foo.sh) in the "Startup command" I have collected this information with `pip list --verbose`

```
2022-05-28T11:45:28.119153575Z anyio              3.6.1       /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.119162175Z asgiref            3.5.2       /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.119552869Z certifi            2022.5.18.1 /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.119820365Z charset-normalizer 2.0.12      /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.119827965Z click              8.1.3       /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.120031262Z colorama           0.4.4       /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.120449156Z cowsay             4.0         /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.120458055Z debugpy            1.5.1       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.120461355Z distlib            0.3.4       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.176880297Z filelock           3.4.2       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.176894696Z Flask              2.0.2       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.177264791Z gunicorn           20.1.0      /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.177561586Z h11                0.13.0      /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.177786683Z httptools          0.4.0       /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.177999280Z idna               3.3         /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.178361674Z itsdangerous       2.0.1       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.179536556Z Jinja2             3.0.3       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.186913844Z MarkupSafe         2.0.1       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.187529034Z objprint           0.2.0       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.187537434Z orjson             3.6.5       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.187541434Z pip                21.2.3      /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.187707532Z platformdirs       2.4.1       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.187958428Z psutil             5.8.0       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.188222024Z pydantic           1.9.1       /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.188431521Z python-dotenv      0.20.0      /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.188687517Z PyYAML             6.0         /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.188877714Z setuptools         57.4.0      /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.189111710Z six                1.16.0      /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.189119810Z sniffio            1.2.0       /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.196622596Z starlette          0.19.1      /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.196818793Z subprocess32       3.5.4       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.196829093Z typing_extensions  4.2.0       /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.197030090Z urllib3            1.26.9      /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.198777263Z uvloop             0.16.0      /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.198787463Z virtualenv         20.13.0     /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.198791263Z vizplugins         0.1.2       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.198794263Z viztracer          0.14.3      /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.198797363Z watchgod           0.8.2       /tmp/8da409e7869f0af/antenv/lib/python3.9/site-packages pip
2022-05-28T11:45:28.198800363Z Werkzeug           2.0.2       /opt/python/3.9.7/lib/python3.9/site-packages           pip
2022-05-28T11:45:28.198803263Z wheel              0.37.0      /opt/python/3.9.7/lib/python3.9/site-packages           pip
```

As you can see, neither FastAPI nor uvicorn are there. It's clear that they are not in the virtual environment.

My pipeline is also very simple (I think).

```yaml
name: Build and deploy Python app to Azure Web App - app-27052022

on:
  push:
    branches:
      - main
  workflow_dispatch:

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Set up Python version
        uses: actions/setup-python@v1
        with:
          python-version: '3.9'

      - name: Create and start virtual environment
        run: |
          python -m pip install --upgrade pip
          pip install pipenv
          pipenv sync --dev        

      - name: Create requirements.txt
        run: |          
          pipenv requirements > ./src/a_project/requirements.txt
      
      - name: Upload artifact for deployment jobs
        uses: actions/upload-artifact@v2
        with:
          name: python-app
          path: |
            ./src/a_project

  deploy:
    runs-on: ubuntu-latest
    needs: build
    environment:
      name: 'Production'
      url: ${{ steps.deploy-to-webapp.outputs.webapp-url }}

    steps:
      - name: Download artifact from build job
        uses: actions/download-artifact@v2
        with:
          name: python-app
          path: .
          
      - name: 'Deploy to Azure Web App'
        uses: azure/webapps-deploy@v2
        id: deploy-to-webapp
        with:
          app-name: 'app-27052022'
          slot-name: 'Production'
          publish-profile: ${{ secrets.AZUREAPPSERVICE_PUBLISHPROFILE_55B3557014E8471B90027396712B1147 }}
```

Basically, I don't know why FastAPI and uvicorn are missed during the deployment process. 

Thank you so much.

**Comment by qianz2** [comment]
Hi @[panicoenlaxbox](https://github.com/panicoenlaxbox), thanks for the feedbacks. It seems that both tools are not installed during activating virtual environment process. I'm not sure if **FastAPI** and **uvicorn** are basic tools that used frequently by most of python apps or they're replaceable with any of tools that already installed in the above list? Are you able to try install it using a `PRE_RUN_COMMAND`?

**Comment by BharathiUpputella** [comment]
Hi @qianz2 

We are also facing similar issue.

In our local, using container inside virtual env we are able to test and deploy our code locally and its working fine with uvicorn.

In Azure, when we are trying to deploy , its failing and error says "uvicorn  command not found".
I have tried connecting through bash to app service and there it shows uvicorn version but not sure why application unable to take it from start up command . Start up command: python -m uvicorn main:app.

Your help will be much appreciated.

Thanks in Advance.

**Comment by panicoenlaxbox** [comment]
Hello, Finally, today I have been able to publish successfully a FastAPI application. 

I think that the problem was that I didn't understand well how to fix all the pieces together. 

FastAPI recommends using, not only uvicorn but also Gunicorn. 

So, with this startup command in App Services, I solved all my problems. 
`gunicorn -w 4 -k uvicorn.workers.UvicornWorker main:app`. Moreover, it's the recommend way to publish in the [FastAPI docs](https://fastapi.tiangolo.com/deployment/server-workers/). 

This link gave me also some clues https://stackoverflow.com/questions/66362199/what-is-the-difference-between-uvicorn-and-gunicornuvicorn 

In summary, if you are going to use App Services to publish your FastAPI application, you will end with the following components:

- nginx (web server/reserve proxy)
- Gunicorn (WSGI server)
- uvicorn (ASGI server)
- FastAPI (your ASGI compliant app)

And today, I published (with az webapp deploy and also with a Github action) and I didn't have problems with dependencies. 

Although process publication seems to behave very odd https://github.com/projectkudu/kudu/issues/3042

For me, this issue can be closed.

Thank you so much.
